File: frontend/src/types/workflow.ts
export type WorkflowStatus = 'queued' | 'running' | 'success' | 'failed';

export interface WorkflowStepRun {
  order: number;
  step: string;
  substep?: string | null;
  algorithm: string;
  params?: Record<string, any> | null;
  status: WorkflowStatus;
  elapsed_ms?: number | null;
  metrics?: Record<string, any> | null;
  visuals?: Array<Record<string, any>> | null;
  error?: string | null;
}

export interface WorkflowRun {
  id: string;
  dataset_id: number;
  template_id?: string | null;
  status: WorkflowStatus;
  started_at?: string | null;
  finished_at?: string | null;
  steps: WorkflowStepRun[];
}

export interface TemplateStep {
  step: string;
  algorithm: string;
  substep?: string | null;
  params?: Record<string, any> | null;
}

export interface WorkflowTemplate {
  id: string;
  name: string;
  version: number;
  description?: string | null;
  steps: TemplateStep[];
  created_at?: string | null;
}

File: backend/app/schemas/workflow.py
from typing import Any, Dict, List, Optional
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, Field

class TemplateStep(BaseModel):
    tool: Optional[str] = None
    step: str
    algorithm: str
    substep: Optional[str] = None
    params: Optional[Dict[str, Any]] = None

class WorkflowTemplateCreate(BaseModel):
    name: str
    description: Optional[str]
    steps: List[TemplateStep]

class WorkflowTemplateUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    steps: Optional[List[TemplateStep]] = None

class WorkflowRunCreate(BaseModel):
    dataset_id: int
    template_id: Optional[UUID]

class WorkflowStepRunRead(BaseModel):
    order: int
    step: str
    substep: Optional[str]
    algorithm: str
    params: Optional[Dict[str, Any]]
    status: str
    elapsed_ms: Optional[int]
    metrics: Optional[Dict[str, Any]]
    visuals: Optional[List[Dict[str, Any]]]
    error: Optional[str]

    class Config:
        orm_mode = True

class WorkflowRunRead(BaseModel):
    id: UUID
    dataset_id: int
    template_id: Optional[UUID]
    status: str
    started_at: Optional[datetime]
    finished_at: Optional[datetime]
    steps: List[WorkflowStepRunRead] = Field(default_factory=list)

    class Config:
        orm_mode = True

class WorkflowTemplateRead(BaseModel):
    id: UUID
    name: str
    version: int
    description: Optional[str]
    steps: List[TemplateStep]
    created_at: Optional[datetime]

    class Config:
        orm_mode = True
